evalml/pipelines/time_series_regression_pipeline.py

```diff
@@ -0,0 +1,110 @@
+            component_graph (list): List of components in order. Accepts strings or ComponentBase subclasses in the list
+
+        Arguments:
+            parameters (dict): Dictionary with component names as keys and dictionary of that component's parameters as values.
+                 An empty dictionary {} implies using all default values for component parameters. Pipeline-level
+                 parameters such as gap and max_delay must be specified with the "pipeline" key. For example:
+                 Pipeline(parameters={"pipeline": {"max_delay": 4, "gap": 2}}).
+            random_state (int, np.random.RandomState): The random seed/state. Defaults to 0.
+        """
+        if "pipeline" not in parameters:
+            raise ValueError("gap and max_delay parameters cannot be omitted from the parameters dict. "
+                             "Please specify them as a dictionary with the key 'pipeline'.")
+        pipeline_params = parameters["pipeline"]
+        self.gap = pipeline_params['gap']
+        self.max_delay = pipeline_params['max_delay']
+        super().__init__(parameters, random_state)
+
+    def fit(self, X, y):
+        """Fit a time series regression pipeline.
+
+        Arguments:
+            X (ww.DataTable, pd.DataFrame or np.ndarray): The input training data of shape [n_samples, n_features]
+            y (ww.DataColumn, pd.Series, np.ndarray): The target training targets of length [n_samples]
+
+        Returns:
+            self
+        """
+        if X is None:
+            X = pd.DataFrame()
+        X = _convert_to_woodwork_structure(X)
+        y = _convert_to_woodwork_structure(y)
+        X = _convert_woodwork_types_wrapper(X.to_pandas())
+        y = _convert_woodwork_types_wrapper(y.to_pandas())
+
+        X_t = self._compute_features_during_fit(X, y)
+
+        y_shifted = y.shift(-self.gap)
+        X_t, y_shifted = keep_non_nan_rows(X_t, y_shifted)
+        self.estimator.fit(X_t, y_shifted)
+        return self
+
+    def predict(self, X, y=None, objective=None):
+        """Make predictions using selected features.
+
+        Arguments:
+            X (ww.DataTable, pd.DataFrame, or np.ndarray): Data of shape [n_samples, n_features]
+            y (ww.DataColumn, pd.Series, np.ndarray, None): The target training targets of length [n_samples]
+            objective (Object or string): The objective to use to make predictions
+
+        Returns:
+            pd.Series: Predicted values.
+        """
+        if X is None:
+            X = pd.DataFrame()
+        X = _convert_to_woodwork_structure(X)
+        y = _convert_to_woodwork_structure(y)
+        X = _convert_woodwork_types_wrapper(X.to_pandas())
+        y = _convert_woodwork_types_wrapper(y.to_pandas())
+
+        features = self.compute_estimator_features(X, y)
+        predictions = self.estimator.predict(features.dropna(axis=0, how="any"))
+        return pad_with_nans(predictions, self.max_delay if features.isna().values.any() else 0)
```

Comment:
@freddyaboulton this looks good!

Why not say `pad_with_nans(predictions, self.max_delay)` ? I figure we should rely on the configuration rather than on properties of the data. Fundamentally, we need the shape of the output to match the number of rows in X and y.

Reply:
If the user doesn't have any delayed features (either the `DelayedFeatureTransformer` is missing from the component graph or they passed `delay_features:False, 'delay_target:False` to `DelayFeatureTransformer`) then we don't need to pad the predictions.

`pad_with_nans(predictions, self.max_delay)` will always add `max_delay` number of rows with NaNs to the beginning.

Reply:
Got it, thanks.

Edge cases to consider:
* `delay_features` and `delay_target` were `False`. And, due to a bug or unexpected component behavior, `features` happens to have a `nan` value somewhere, even though there were no `nans` created by the delayed featurizer.
* `delay_features` or `delay_target` were `True`. And, due to a bug or unexpected component behavior, `features` happens to not have any `nans`. Maybe this case isn't important, but it could happen if someone put an imputer after the delayed feature eng.

Is there a foolproof way to avoid these sorts of scenarios? What about
```python
n_rows_padding_pre = X.shape[0] - predictions.shape[0]
return pad_with_nans(predictions, n_rows_padding_pre)
```

Reply:
Yea I think that works! Good suggestion.